### PR TITLE
Support unbound generic types in 'nameof' operator.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Championed proposal: https://github.com/dotnet/csharplang/issues/8480

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 1]
```diff
@@ -1212,7 +1212,7 @@ private TypeWithAnnotations BindGenericSimpleNamespaceOrTypeOrAliasSymbol(
 
             if (isUnboundTypeExpr)
             {
-                if (!IsUnboundTypeAllowed(node))
+                if (!IsInsideNameof && !IsUnboundTypeAllowed(node))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1215]: From my understanding, we want `nameof` to behave just like `typeof` with regards to open types.
So it feels like the change should be inside an override of `IsUnboundTypeAllowed` in `NameofBinder` instead, implemented in a similar fashion to `TypeofBinder.IsUnboundTypeAllowed`.

Then depending on where the LangVersion lands, we may not need any change to `ConstructNamedTypeUnlessTypeArgumentOmitted` below and we may not need dedicated validation logic (`validateOpenGenericNames` and `ERR_NestedUnboundTypeNotAllowedInNameofExpression`).

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 2]
```diff
@@ -2373,5 +2366,334 @@ class Attr : System.Attribute { public Attr(string s) {} }";
...
+        [Fact]
+        public void OpenTypeInNameof_CSharp13()
+        {
+            CreateCompilation("""
+                using System;
+                using System.Collections.Generic;
+
+                var v = nameof(List<>);
+                Console.WriteLine(v);
+                """, parseOptions: TestOptions.Regular13).VerifyDiagnostics(
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2379]: >Regular13 [](http://example.com/codeflow?start=47&length=9)

Tests mean to test specific language version should also include RegularNext (ie. means Regular14) and Preview.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 3]
```diff
@@ -2373,5 +2366,334 @@ class Attr : System.Attribute { public Attr(string s) {} }";
...
+        {
+            CompileAndVerify(
+                CreateCompilation("""
+                    using System;
+                    
+                    var v = nameof(A<>.B<int>);
+                    Console.WriteLine(v);
+
+                    class A<X> { public class B<Y>; }
+                    """, parseOptions: TestOptions.RegularPreview).VerifyDiagnostics(),
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2461]: >parseOptions: TestOptions.RegularPreview [](http://example.com/codeflow?start=25&length=40)

nit: `RegularPreview` is the default for `CreateCompilation` so can be omitted

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -6027,6 +6027,15 @@ private ScanTypeFlags ScanPossibleTypeArgumentList(
                     return result;
                 }
 
+                // Allow for any chain of errant commas in teh generic name.  like `Dictionary<,int>` or
+                // `Dictionary<int,,>` We still want to think of these as generics, just with missing type-arguments, vs
+                // some invalid tree-expression that we would otherwise form.
+                if (this.CurrentToken.Kind == SyntaxKind.CommaToken)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:6033]: nit: consider adding a parsing test showing the effect of the change

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 5]
```diff
@@ -2373,5 +2366,334 @@ class Attr : System.Attribute { public Attr(string s) {} }";
...
+                    // (3,20): error CS8652: The feature 'Unbound generic types in nameof operator' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                    // var v = nameof(A<>.B<>);
+                    Diagnostic(ErrorCode.ERR_FeatureInPreview, "B<>").WithArguments("Unbound generic types in nameof operator").WithLocation(3, 20));
+        }
+
+        [Fact]
+        public void OpenTypeInNameof_BaseCase()
+        {
+            CompileAndVerify(
+                CreateCompilation("""
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2440]: nit: This and similar tests could be simplified by removing `CreateCompilation`: 
`CompileAndVerify("... source ..., expectedOutput: "...").VerifyDiagnostics();`

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 17)

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs [hunk 6]
```diff
@@ -294,6 +294,7 @@ internal enum MessageID
         IDS_OverloadResolutionPriority = MessageBase + 12848,
 
         IDS_FeatureFirstClassSpan = MessageBase + 12849,
+        IDS_FeatureUnboundGenericTypesInNameof = MessageBase + 12850,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Errors/MessageID.cs:297]: Some paperwork: You'll want to add an entry at the top of "Working Set C#" section on https://github.com/dotnet/roslyn/blob/main/docs/Language%20Feature%20Status.md

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Errors/MessageID.cs:297]: More paperwork :-P You'll want to convert the championed proposal to new format (ie. move the spec portion into a speclet document in `proposals` folder)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2379]: done!

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: Done!

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:6058]: Tests added.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2464]: Done.  Thanks!

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 7]
```diff
@@ -2221,6 +2221,7 @@ private BoundExpression BindNameofOperatorInternal(InvocationExpressionSyntax no
         {
             CheckFeatureAvailability(node, MessageID.IDS_FeatureNameof, diagnostics);
             var argument = node.ArgumentList.Arguments[0].Expression;
+
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:2224]:
```suggestion
```

File: src/Compilers/CSharp/Portable/Binder/NameofBinder.cs [hunk 8]
```diff
@@ -19,19 +21,25 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// </summary>
     internal sealed class NameofBinder : Binder
     {
-        private readonly SyntaxNode _nameofArgument;
+        private readonly ExpressionSyntax _nameofArgument;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/NameofBinder.cs:24]: This always had to be an ExpressionSyntax.  Making strongly typed and to match the sig of OpenTypeVisitor.VisitThis

File: src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    /// 
+    /// Open types are allowed if an only if:
+    /// <list type="number">
+    /// <item>There is no constructed generic type elsewhere in the visited syntax; and</item>
+    /// <item>The open type is not used as a type argument or array/pointer/nullable element type.</item>
+    /// </list>
+    /// 
+    /// Open types can be used both in <c>typeof(...)</c> and <c>nameof(...)</c> expressions.
+    /// </summary>
+    protected sealed class OpenTypeVisitor : CSharpSyntaxVisitor
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs:23]: extracted out of TypeOfBinder so it can be used by TypeOfBinder and NameOfBinder.

File: src/Compilers/CSharp/Portable/Binder/TypeofBinder.cs [hunk 10]
```diff
@@ -20,7 +20,7 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// should be considered open so that the flag can be set 
     /// appropriately in BoundTypeOfOperator.
     /// </summary>
-    internal sealed class TypeofBinder : Binder
+    internal sealed partial class TypeofBinder : Binder
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/TypeofBinder.cs:23]:
```suggestion
    internal sealed class TypeofBinder : Binder
```

File: src/Compilers/CSharp/Portable/Binder/TypeofBinder.cs [hunk 11]
```diff
@@ -32,96 +32,6 @@ internal TypeofBinder(ExpressionSyntax typeExpression, Binder next)
...
-        }
-
-        /// <summary>
-        /// This visitor walks over a type expression looking for open types.
-        /// Open types are allowed if an only if:
-        ///   1) There is no constructed generic type elsewhere in the visited syntax; and
-        ///   2) The open type is not used as a type argument or array/pointer/nullable
-        ///        element type.
-        /// </summary>
-        private class OpenTypeVisitor : CSharpSyntaxVisitor
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/TypeofBinder.cs:47]: moved to sibling type so it could be shared.

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Errors/MessageID.cs:297]: will do :)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 12]
```diff
@@ -2373,5 +2366,497 @@ class Attr : System.Attribute { public Attr(string s) {} }";
...
+            var root = tree.GetRoot();
+
+            var firstGeneric = root.DescendantNodes().OfType<GenericNameSyntax>().First();
+            var lastGeneric = root.DescendantNodes().OfType<GenericNameSyntax>().Last();
+
+            Assert.NotSame(firstGeneric, lastGeneric);
+
+            // Ensure the type inside the nameof is the same as the type inside the typeof.
+            var type1 = semanticModel.GetTypeInfo(firstGeneric).Type;
+            var type2 = semanticModel.GetTypeInfo(firstGeneric).Type;
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2854]: >firstGeneric [](http://example.com/codeflow?start=50&length=12)

Crucial typo ;-)
Once fixed, this will show that one will be a definition and one will be substituted...

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 13]
```diff
@@ -2373,5 +2366,497 @@ class Attr : System.Attribute { public Attr(string s) {} }";
...
+        }
+
+        [Fact]
+        public void Nameof_NestedClosedType2()
+        {
+            CreateCompilation("""
+                using System;
+                using System.Collections.Generic;
+
+                var v = nameof(List<List<int>>);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2642]: I couldn't see the difference with `Nameof_NestedClosedType1`

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 34)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2854]: good catch.

**Comment by CyrusNajmabadi** [comment]
> From my understanding, we want nameof to behave just like typeof with regards to open types.

Sort of...

Except a few things:

1. typeof is a type only context, and cannot ever reference members within it.
2. nameof allows member lookup.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs (hunk 13 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: fixed.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 14]
```diff
@@ -1373,22 +1373,38 @@ private NamedTypeSymbol ConstructNamedTypeUnlessTypeArgumentOmitted(SyntaxNode t
...
+                    CheckFeatureAvailability(typeSyntax, MessageID.IDS_FeatureUnboundGenericTypesInNameof, diagnostics);
+
+                    // From the spec:
+                    //
+                    // Member lookup on an unbound type expression will be performed the same way as for a `this`
+                    // expression within that type declaration.
+                    //
+                    // So we want to just return the originating type symbol as is (e.g. List<T> in nameof(List<>)).
+                    // This is distinctly different than how typeof(List<>) works, where it returns an unbound generic
+                    // type.
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1388]: For the question of return a definition vs. "unbound" (`SubstitutedNamedTypeSymbol._unbound`) and the behavior of GetMembers, I think this needs a bit more investigation.
I like what you have in the spec and the we've aligned the implementation to return the definition, but something feels off. Why didn't we return the definition for unbound in other scenarios? (why did we need the `_unbound` flag?)
It's not clear to me whether the GetMembers behavior that you pointed out is a feature (we really don't want to return non-type members from unbound because of reason X), or an optimization (we don't need to bother because no scenario needs non-type members on unbound types).

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1388]: Who does this investigation?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1388]: Discussed with @AlekseyTs and team at standup today. 
TL;DR: returning the definition (rather than an "unbound" symbol) is fine :-)

For purpose of `typeof` we need to differentiate between an unbound type and the definition, because it would impact what IL we emit. Also as we'd discussed, for `typeof` we ever only needed nested types (not other kinds of members).  
But for purpose of `nameof` we don't need that, the type definition is fine, since we'll only need its name and we do want both type and non-type members.  
This choice will be visible in the symbol from the semantic model, but the difference is minor.

Also, FWIW, the [standard](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/types.md#844-bound-and-unbound-types) supports using the definition for unbound types in general:
> The term unbound type refers to a non-generic type or an unbound generic type. The term bound type refers to a non-generic type or a constructed type.
> An unbound type refers to the entity declared by a type declaration. An unbound generic type is not itself a type, and cannot be used as the type of a variable, argument or return value, or as a base type.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 15]
```diff
@@ -2373,5 +2366,507 @@ class Attr : System.Attribute { public Attr(string s) {} }";
...
+                using System;
+                    
+                var v1 = nameof({{nameofTypeString}});
+                var v2 = typeof(IGoo<>);
+                Console.WriteLine(v1 + v2);
+
+                interface IGoo<T> { public T Count { get; } }
+                """).VerifyDiagnostics();
+            var tree = compilation.SyntaxTrees.Single();
+            var semanticModel = compilation.GetSemanticModel(tree);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2843]: It would be good to explicitly check `IsDefinition` since that's the most directly observable difference between those two symbols.
Also, consider testing `IsUnboundGenericType`.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 38)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs (hunk 15 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2843]: done.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler  for another set of eyes.  Tnx :)

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler for another set of eyes. Tnx :)

**Comment by AlekseyTs** [comment]
@CyrusNajmabadi I think it would be good to have a speclet for the feature.

**Comment by CyrusNajmabadi** [comment]
Spec is here: https://github.com/dotnet/csharplang/blob/main/proposals%2Funbound-generic-types-in-nameof.md

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 16]
```diff
@@ -4932,6 +4932,9 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="IDS_FeatureNameof" xml:space="preserve">
     <value>nameof operator</value>
   </data>
+  <data name="IDS_FeatureUnboundGenericTypesInNameof" xml:space="preserve">
+    <value>Unbound generic types in nameof operator</value>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:4936]: >Unbound [](http://example.com/codeflow?start=11&length=7)

Usually feature strings do not start with a capital letter

**Comment by AlekseyTs** [comment]
It feels like we would want to have similar relaxation in VB

File: src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+            visitor.Visit(typeSyntax);
+            allowedMap = visitor._allowedMap;
+        }
+
+        public override void VisitGenericName(GenericNameSyntax node)
+        {
+            SeparatedSyntaxList<TypeSyntax> typeArguments = node.TypeArgumentList.Arguments;
+            if (node.IsUnboundGenericName)
+            {
+                _allowedMap ??= [];
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs:45]: >[] [](http://example.com/codeflow?start=32&length=2)

Please keep the original style

File: src/Compilers/CSharp/Portable/Binder/TypeofBinder.cs [hunk 18]
```diff
@@ -32,96 +32,6 @@ internal TypeofBinder(ExpressionSyntax typeExpression, Binder next)
...
-                Visit(node.ElementType);
-            }
-
-            public override void VisitNullableType(NullableTypeSyntax node)
-            {
-                _seenConstructed = true;
-                Visit(node.ElementType);
-            }
-        }
+            => _allowedMap != null && _allowedMap.TryGetValue(syntax, out bool allowed) && allowed;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/TypeofBinder.cs:35]: >=> _allowedMap != null && _allowedMap.TryGetValue(syntax, out bool allowed) && allowed; [](http://example.com/codeflow?start=12&length=87)

Unless there is a desire to change behavior of the function, please revert the change

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/TypeofBinder.cs:35]: >=> [](http://example.com/codeflow?start=12&length=2)

Please keep original body style.

File: src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+            Visit(node.Name);
+        }
+
+        public override void VisitArrayType(ArrayTypeSyntax node)
+        {
+            _seenConstructed = true;
+            Visit(node.ElementType);
+        }
+
+        public override void VisitPointerType(PointerTypeSyntax node)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs:87]: >VisitPointerType [](http://example.com/codeflow?start=29&length=16)

Should ```VisitFunctionPointerType``` be overridden as well?

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 16 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/CSharpResources.resx]: sgtm.

**Comment by CyrusNajmabadi** [comment]
> It feels like we would want to have similar relaxation in VB

Seems like it would be nice there as well.  If someone wants to champion that, i'd support it.

File: src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs (hunk 19 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs:87]: Not sure.  @333fred do function pointers need special handling here?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs:87]: note: i've added some typeof and nameof tests taht exercise func-ptrs.  The results seem expected to me.  But i wouldn't mind SMEs here validating.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 20]
```diff
@@ -1373,22 +1373,38 @@ private NamedTypeSymbol ConstructNamedTypeUnlessTypeArgumentOmitted(SyntaxNode t
...
+                    Error(diagnostics, ErrorCode.ERR_BadArity, typeSyntax, type, MessageID.IDS_SK_TYPE.Localize(), typeArgumentsSyntax.Count);
+                }
 
-                // If the syntax looks like an unbound generic type, then they probably wanted the definition.
-                // Give an error indicating that the syntax is incorrect and then use the definition.
-                // CONSIDER: we could construct an unbound generic type symbol, but that would probably be confusing
-                // outside a typeof.
                 return type;
             }
-            else
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1386]: >else [](http://example.com/codeflow?start=12&length=4)

Please restore original style

**Comment by AlekseyTs** [comment]
> > It feels like we would want to have similar relaxation in VB
> 
> Seems like it would be nice there as well. If someone wants to champion that, i'd support it.

If you are not planning to adjust VB, please open an issue to follow up and CC @KathleenDollard on it.

**Comment by CyrusNajmabadi** [comment]
Opened https://github.com/dotnet/roslyn/issues/75975

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 21]
```diff
@@ -2373,5 +2366,687 @@ class Attr : System.Attribute { public Attr(string s) {} }";
...
+        [Fact]
+        public void OpenTypeInNameof_Preview()
+        {
+            CreateCompilation("""
+                using System;
+                using System.Collections.Generic;
+
+                var v = nameof(List<>);
+                Console.WriteLine(v);
+                """).VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs:2379]: >VerifyDiagnostics(); [](http://example.com/codeflow?start=21&length=20)

Let's also execute the code and observe behavior at runtime. This is probably applicable to other unit-tests targeting success scenarios

File: src/Compilers/Test/Utilities/CSharp/Extensions.cs [hunk 22]
```diff
@@ -764,9 +764,7 @@ public static ImmutableArray<IParameterSymbol> GetParameters(this ISymbol member
     }
 
     public static bool IsUnboundGenericType(this ITypeSymbol type)
-    {
-        return type is INamedTypeSymbol namedType && namedType.IsUnboundGenericType;
-    }
+        => type is INamedTypeSymbol { IsUnboundGenericType: true };
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp/Extensions.cs:767]: >=> [](http://example.com/codeflow?start=8&length=2)

Please restore original style

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 49)

**Comment by AlekseyTs** [comment]
@CyrusNajmabadi Please squash commit while merging.

File: src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs (hunk 19 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs:87]: I've played with the change a bit and looked through the logic, and I can't see a scenario where it should matter; since we use `DefaultVisit` for the function pointer, no nested generic types will be marked as allowed. And for any containers, it seems that the following should always hold:

* The container is a generic type, and we won't parse that as an unbound generic type if there are any arguments (ie, `IsUnboundGenericName` won't be true for something like `Dictionary<,delegate*<List<>>`)
* The container is an array or nullable type, and that's always marked as seen constructed.
* The container is a tuple type, and that's never visited by this visitor, so nothing underneath is allowed to be open.

I think that covers all the possibilities, so I don't expect there to be any issues. However, @CyrusNajmabadi, please add a test with 3 generics used like this:
`C<, delegate*<int>, List<>>`

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/OpenTypeVisitor.cs:87]: @333fred tests added.  test results seemed sensible to me, so i'm moving forward with submitting this.  tnx.
